### PR TITLE
Validate user task assignments

### LINKED ISSUE #8054: Validate assignment definitions
**Description**

Processes are validated for model correctness when they are deployed by a user. A new validation is needed to make sure assignment definitions (from #8053) for a user task are provided correctly.

- validation failures should behave like any other model validation failure.
- the values of the `assignee` and `candidateGroups` must be parsed as expressions.
- an additional transformation is required for static value definitions of the `candidateGroups` (i.e. from comma-separated list of values to FEEL-formatted list of string).
- in case of static expressions, we need to additionally check the formatting of the parsed static expressions.
- the parsed expressions should be stored in the executable model so they can be evaluated at runtime

See #8053 for the specs.

### PR DESCRIPTION
**Pull request by korthout**
## Description

Adds runtime validations to make sure that user task assignments comply with the requirements.

Specifically, it verifies the `assignee` and `candidateGroups` attributes of the `zeebe:AssignmentDefinition` extension element. Both attributes are optional, and can be either a static value or a FEEL expression. For candidate groups special rules apply when it is a static value.

No validation is added to make sure either attributes is provided on an assignment definition element. Additional care must be taken for this when implementing #8055.

## Related issues

closes #8054 

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [x] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [x] There are unit/integration tests that verify all acceptance criterias of the issue
* [x] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

**Review by saig0 — COMMENTED** [review]
@korthout looks good :+1: 

But I have a question regarding the `candidateGroups`. Please have a look.

File: engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/validation/ZeebeRuntimeValidationTest.java [hunk 1]
```diff
@@ -332,11 +332,38 @@
...
+      {
+        /* invalid candidateGroups static value */
+        Bpmn.createExecutableProcess("process")
+            .startEvent()
+            .userTask("task", b -> b.zeebeCandidateGroups("a,b,c"))
+            .done(),
+        List.of(
+            expect(
+                ZeebeAssignmentDefinition.class,
+                "Expected static value to be a bracketed list of comma-separated values, e.g. '[a,b,c]', but found 'a,b,c'"))
```
[the remark this answers is not in the document]
Comment (korthout) [reply engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/validation/ZeebeRuntimeValidationTest.java:365]: Oops! No this was my bad. Glad we have a review to spot stuff like this. Thanks @saig0 💯

**Comment by korthout** [comment]
@saig0 Please have another look. I'll rebase (autosquash) the fixup commits into the referred commits after approval, but left them here for easier review.

File: engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/ZeebeExpressionValidator.java [hunk 2]
```diff
@@ -99,6 +103,19 @@ private static void verifyPath(
...
   }
 
+  public static boolean isListOfCsv(final Expression staticExp) {
+    final String value = staticExp.getExpression();
+    if (value.isEmpty()) {
+      return true;
+    }
+    final var values = value.split(",");
+    if (values.length < StringUtils.countMatches(value, ",") + 1) {
+      // one of the values was an empty string, e.g. [a,,c]
```
Comment (saig0) [inline engine/src/main/java/io/camunda/zeebe/engine/processing/deployment/model/validation/ZeebeExpressionValidator.java:113]: :wrench: Fix comment.

```suggestion
      // one of the values was an empty string, e.g. 'a,,c'
```

File: engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/validation/ZeebeExpressionValidatorTest.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+class ZeebeExpressionValidatorTest {
+
+  @Nested
+  @DisplayName("ZeebeExpressionValidator.isListOfCsv(Expression)")
+  @TestInstance(TestInstance.Lifecycle.PER_CLASS)
+  class IsBracketedListOfValues {
```
Comment (saig0) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/validation/ZeebeExpressionValidatorTest.java:26]: :wrench: We could rename the inner class.

```suggestion
  class IsListOfValues {
```

File: engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/validation/ZeebeExpressionValidatorTest.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+          Arguments.of("a"),
+          Arguments.of("a,b"),
+          Arguments.of("a,b,c"),
+          Arguments.of("\"abcd\",\"efgh\",\"ijkl\""),
+          Arguments.of("1"),
+          Arguments.of("1,2"),
+          Arguments.of("1,a,3"),
+          Arguments.of("[]"),
+          Arguments.of("[],[]"),
+          Arguments.of("[,][,]"));
```
Comment (saig0) [inline engine/src/test/java/io/camunda/zeebe/engine/processing/deployment/model/validation/ZeebeExpressionValidatorTest.java:54]: :thinking: These cases are uncommon for a CSV but okay.

**Review by saig0 — APPROVED** [review]
@korthout thanks :+1:  I see just two minor clean ups :rocket: 

> I'm unsure where I got in my head that candidate groups as static value
should be a bracketed list of csv, but that was wrong.

I like the commit description :laughing:

**Pull request merged**